### PR TITLE
Bv sync

### PR DESCRIPTION
**Pull request by ndcgundlach**
## Change Description

Added calls to blockvault to perform syncing and handle snapshots and block from blockvault.
## Change Type
**Select ONE**
- [ ] Documentation

- [ ] Stability bug fix

- [ ] Other

- [ ] Other - special case

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: plugins/chain_plugin/chain_plugin.cpp [hunk 1]
```diff
@@ -184,6 +184,7 @@ class chain_plugin_impl {
    bool                             accept_transactions = false;
    bool                             api_accept_transactions = true;
    bool                             account_queries_enabled = false;
+   bool                             blockvault_enabled = false;
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp:187]: Not used.

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            blockvault_sync_strategy(block_vault_interface* blockvault, BP& blockchain_provider, std::function<void()> shutdown,
+                                     std::function<bool()> check_shutdown, chain::block_id_type block_id) :
+                _blockvault(blockvault), _blockchain_provider(blockchain_provider), _shutdown(shutdown), _check_shutdown(check_shutdown),
+                _previous_block_id(block_id) {
+            }
+
+            void do_sync() {
+                std::string previous_block_id;
+                if (_previous_block_id.has_value()) {
+                    previous_block_id = _previous_block_id->str();
```
Comment (heifner) [inline plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:48]: Should not call `->str()`

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                }
+                _blockvault->sync(previous_block_id, *this);
```
Comment (heifner) [inline plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:50]: Why does sync take block id as a string?

Comment (ndcgundlach) [reply plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:50]: This was the format that @huangminghuang specified.

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            }
+
+            void on_snapshot(std::string_view snapshot_filename) override {
```
Comment (johndebord) [inline plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:53]: Should `override` be `final` if `blockvault_sync_strategy` doesn't plan on being a paren't class to anything in the future?

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                auto infile = std::ifstream(snapshot_filename, (std::ios::in | std::ios::binary));
+                auto reader = std::make_shared<chain::istream_snapshot_reader>(infile);
+                _blockchain_provider.chain->startup(_shutdown, _check_shutdown, reader);
+                infile.close();
+
+                _got_snapshot = true;
+            }
+
+            void on_block(std::string_view block) override {
```
Comment (johndebord) [inline plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:62]: Should `override` be `final` if `blockvault_sync_strategy` doesn't plan on being a paren't class to anything in the future?

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
+#pragma once
+#include <stdint.h>
+#include <string_view>
```
Comment (johndebord) [inline plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:3]: Tested that it builds without including `#include <stdint.h>` and `#include <string_view>`

File: plugins/chain_plugin/chain_plugin.cpp [hunk 7]
```diff
@@ -1307,7 +1308,13 @@ void chain_plugin::plugin_startup()
    try {
       auto shutdown = [](){ return app().quit(); };
       auto check_shutdown = [](){ return app().is_quiting(); };
-      if (my->snapshot_path) {
+      auto blockvault_instance = eosio::blockvault_client_plugin::get();
+      if (nullptr != blockvault_instance) {
+          auto head_block_num = my->chain->head_block_id();
```
Comment (heifner) [inline plugins/chain_plugin/chain_plugin.cpp:1313]: Rename `head_block_num` -> `head_block_id`

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                auto infile = std::ifstream(snapshot_filename, (std::ios::in | std::ios::binary));
+                auto reader = std::make_shared<chain::istream_snapshot_reader>(infile);
+                _blockchain_provider.chain->startup(_shutdown, _check_shutdown, reader);
+                infile.close();
+
+                _got_snapshot = true;
+            }
+
+            void on_block(eosio::chain::signed_block_ptr block) override final {
+                if (!_got_snapshot) {
```
Comment (heifner) [inline plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp:39]: Undefined behavior if this is called before `on_snapshot`. `_got_snapshot` is not initialized.

File: plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp (hunk 3 above)
Comment (heifner) [reply plugins/chain_plugin/include/eosio/chain_plugin/blockvault_sync_strategy.hpp]: Ah because a string is stored in the db. That is fine.

**Review by heifner — APPROVED** [review]

**Pull request merged**